### PR TITLE
[WIP] Ops in Python

### PR DESCRIPTION
**Pull request by mruberry**
Draft only.

Latest draft file to discuss is tools/codegen/pyops/gen_cpp.py, other files as they were originally for history.

The updated draft file demonstrates the (very) simple transpiler IR, how type propagation works, how calls can be easily remapped based on the type they're called on, how torch functions can be understood by reading native_functions.yaml, and how a simple pass that auto-generates a function's out= variant work. 

The draft takes the original dstack Python:

```
def dstack(tensors: TensorList) -> Tensor:
    TORCH_CHECK(len(tensors) > 0, "dstack expects a non-empty list of tensors")
    rep = torch.atleast_3d(tensors)
    return torch.cat(rep, 2)
```

And generates the following sample body C++ (minus semicolons and without remapping the out= call for simplicity):

```
auto v1 = tensors.size()  // "len" is treated as a built-in that can be applied to the TensorList type as a size member access
auto v2 = (v1 > 0)
TORCH_CHECK(v2, 'dstack expects a non-empty list of tensors')  // even though this produces code in SSA format literals are inlined for readability
auto rep = at::atleast_3d(tensors)
auto v5 = at::cat(rep, 2)
return v5

// out variant
auto v1 = tensors.size()
auto v2 = (v1 > 0)
TORCH_CHECK(v2, 'dstack expects a non-empty list of tensors')
auto rep = at::atleast_3d(tensors)
auto v5 = at::cat(rep, 2, out=result) // note this call still appears "Pythonically" for ease of demo implementation
```

**Proposed TODOs to get out of draft**

- finish calling convention
- tests of Python vs. C++ correctness
- tests of the transpiler itself
- strict mypy the (two) IR nodes
- documentation
- if statements?
- additional operators to demonstrate coverage/naturalness
  - hstack, vstack
  - atleast_1d, atleast_2d, atleast_3d requires multiple signatures and conditionals, plus would be a PyOp depended on by a PyOp
  - rot90 requires mathematical expressions and inplace operations
  - bartlett_window requires handling multiple signatures, TensorOptions, casts, inplace operations, and mathematical operations
  - tensor_split requires handling multiple signatures, mathematical operations, for loops, and tensor metadata
- @ngimel implements 1+ operators to validate "naturalness" of the system 
- file modularity

### CODE REVIEW

**Comment by ezyang** [comment]
Nice! Post the generated C++!

File: torch/pyops/__init__.py [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+    except IOError:
+        old_contents = None
+    if contents != old_contents:
+        with open(filename, 'w') as f:
+            f.write(contents)
+
+def TORCH_CHECK(cond, str):
+    assert cond, str
+
+class TensorList(object):
```
Comment (ezyang) [inline torch/pyops/__init__.py:32]: List[Tensor]?

**Comment by ezyang** [comment]
This is a good conversation starter. A lot of details that need more design fleshed out. We can talk about it more later today.

**Comment by mruberry** [comment]
> Nice! Post the generated C++!

Good point. This is the PyOps.cpp file in aten/src/ATen/native that it generates:

```

#include <ATen/ATen.h>
#include <c10/util/Exception.h>

// NOTICE! This file is autogenerated!

namespace at { namespace native {

Tensor dstack(TensorList tensors) {
	TORCH_CHECK(tensors.size() > 0, "dstack expects a non-empty list of tensors");
	auto rep = at::atleast_3d(tensors);
	return at::cat(rep, 2);
}

Tensor& dstack_out(TensorList tensors, Tensor& result) {
	TORCH_CHECK(tensors.size() > 0, "dstack expects a non-empty list of tensors");
	auto rep = at::atleast_3d(tensors);
	return at::cat_out(result, rep, 2);
}

}} // at::native
```

**Comment by ngimel** [comment]
But does it really generate the implementation, or does it (for now) copy the template from https://github.com/pytorch/pytorch/pull/63643/files#diff-242461627b9981321d9d986a7d16f960d89e966d7896b08b1d2b0ab751bcbd2aR285?

**Comment by mruberry** [comment]
> But does it really generate the implementation, or does it (for now) copy the template from https://github.com/pytorch/pytorch/pull/63643/files#diff-242461627b9981321d9d986a7d16f960d89e966d7896b08b1d2b0ab751bcbd2aR285?

Good question because that is confusing. It does really generate the implementation. I was using that template to verify the e2e flow before generation and kept it in as a reference for our discussion.

File: torch/pyops/gen_cpp.py [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+            s = str(self.value) + "." + self.attr
+
+        if type(self.ctx) == ast.Load:
+            s += "()"
+
+        return s
+
+# A function call
+# https://greentreesnakes.readthedocs.io/en/latest/nodes.html?highlight=call#Call
+class _Call(_Statement):
```
Comment (ezyang) [inline torch/pyops/gen_cpp.py:98]: Calls are not statements 🧐

Comment (mruberry) [reply torch/pyops/gen_cpp.py:98]: This is just an old file

**Pull request closed without merging**